### PR TITLE
Api management

### PR DESCRIPTION
**Pull request by mxxt1**
# Description

This feature prevents users from abusing the API by linking the call count to the user id instead of the api key itself. This allows users to generate a new key if they've lost their existing key without circumventing the quota enforcement system. 

###Redis: 

Refactors the existing API key generation functionality to link api call count to the user id. When a key is generated, the system checks for an existing entry using the user id. If a record exists for that user id, the new hashed key overwrites the existing hashed key, and sends the plain-text api key to the frontend, and stays visible during that session.
 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Change Status

- [ ] Complete, tested, ready to review and merge
- [x] Complete, but not tested (may need new tests)
- [x] Incomplete/work-in-progress, PR is for discussion/feedback

# How Has This Been Tested?

- [ ] Postman
- [ ] Test B

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My code has been reviewed by at least one peer
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] There are no merge conflicts

### CODE REVIEW

File: __tests__/Routes/Routes/Users/userRoles.spec.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            "role": "admin"
+        }
+    }
+
+    describe('Should be able to retrieve user data', () => {
+        test('should be able to retrieve full user schema object from Auth0',
+            () => request(server).post(`/api/users/`).send({ "sub": user.sub })
+                .then(result => {
+                    const data = result.body
+                    // console.log('LOGGING', result, data)
```
Comment (blackmacaroon) [inline __tests__/Routes/Routes/Users/userRoles.spec.js]: so proud of myself for commenting the whole pr, did it on the og repo. 🤦‍♀  
bold addition of extra space, I'm all about aesthetics.

File: README.md [hunk 2]
```diff
@@ -1,3 +1,5 @@
+
+
```
Comment (blackmacaroon) [inline README.md:2]: bold addition of extra space, I'm all about aesthetics.

File: __tests__/Routes/Routes/Users/userRoles.spec.js (hunk 1 above)
Comment (blackmacaroon) [inline __tests__/Routes/Routes/Users/userRoles.spec.js:18]: maybe it's a console log for testing, test it and delete it. maybe it's for future devs, leave a comment about why they'll need to console log.

File: api-key/migrations/20191018164134_create-apiKey-table.js [hunk 3]
```diff
@@ -9,9 +9,22 @@ exports.up = function(knex) {
       .text('user_id')
       .unique()
       .notNullable()
+    //reset_date stores the date as milliseconds as an integer
```
Comment (blackmacaroon) [inline api-key/migrations/20191018164134_create-apiKey-table.js:12]: excellent use of comment to clarify potentially confusing code, future devs thank you

File: api/server.js [hunk 4]
```diff
@@ -22,19 +27,27 @@ const moesifMiddleware = moesifExpress({
   logBody: true
 })
 
-//Server uses middleware to add functionality
-server.use(moesifMiddleware)
 
+// TODO: POSSIBL CODE CLEANUP WITH BELOW IMPORTS, AS THEIR NOT BEING USED.
```
Comment (blackmacaroon) [inline api/server.js:31]: if we're not using it, delete it. Or, as I believe the kids are calling it these days: "yeet"

File: api/server.js [hunk 5]
```diff
@@ -53,4 +66,8 @@ server.get('/sauti', (req, res) => {
     })
 })
 
+// TODO: CLEAN UP TEMP CODE AFTER FINISHED WRITING MIDDLEWARE FOR API
+const roles = require('../middleware/rules/rules-middleware')
+// roles();
+
```
Comment (blackmacaroon) [inline api/server.js:72]: yeet

File: middleware/api-limiter-middleware.js [hunk 6]
```diff
@@ -1,27 +1,96 @@
...
+        // Create a new key in redis cache
+        client.set(userId, 1)
+        next()
+      }
   } else {
-    client.set(key, 0) // Create a new key in redis cache
     next()
   }
-}
+};
```
Comment (blackmacaroon) [inline middleware/api-limiter-middleware.js:99]: excellent use of comments as notes for future devs

File: middleware/bottle.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import Bottleneck from "bottleneck";
```
Comment (blackmacaroon) [inline middleware/bottle.js:1]: did we end up actually using Bottleneck? If not, we can go ahead and delete this whole thing.

File: middleware/rules/rules-middleware.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+// * DYNAMIC USER GET REQUEST FOR USERS REQUESTING API KEY
+module.exports = async (req, res, next) => {
+  let apiToken
+
+  // * SET TOKEN FOR AUTHORIZED API USAGE
+  request(options, async (error, response, body) => {
+    apiToken = JSON.parse(body).access_token
+
+    // ! LOG TOKEN DATA BELOW
+    // console.log(apiToken)
```
Comment (blackmacaroon) [inline middleware/rules/rules-middleware.js:29]: for testing or for devs? "no console logs in final code" but I don't hate commented out console logs with explanations in certain situations. Best Practice? Who knows.

File: __tests__/Routes/Routes/Users/userRoles.spec.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+                    expect(!!data.app_metadata.role).toBe(true)
```
Comment (blackmacaroon) [inline __tests__/Routes/Routes/Users/userRoles.spec.js:20]: I had to look up double exclamation, I think I like it, but idk if I get it. It forces a value to behave like a boolean? fancy dev stuff here.

File: middleware/token-middleware.js [hunk 10]
```diff
@@ -1,15 +1,18 @@
...
+      jwksRequestsPerMinute: 10,
+      jwksUri:
+        'https://sauti-africa-market-prices.auth0.com/.well-known/jwks.json'
+    }),
+    audience: 'https://sauti-africa-market-prices.auth0.com/api/v2/',
+    issuer: 'https://sauti-africa-market-prices.auth0.com/',
+    algorithms: ['RS256']
+  })
+  next();
+}
```
Comment (blackmacaroon) [inline middleware/token-middleware.js:18]: If I learned anything in Labs it was: async all the things!

File: redis/index.js [hunk 11]
```diff
@@ -10,6 +10,10 @@ client.on('error', err => {
   console.log('error', err)
 })
 
+
+//this method is no longer applicable with user tiering. Reset periods are now determined by the api key registration date, and calculating 30 day intervals based on that initial generation date
```
Comment (blackmacaroon) [inline redis/index.js:14]: what's no longer applicable?

File: currency/index.js [hunk 12]
```diff
@@ -65,36 +71,69 @@ const getExchangeRates = async () => {
...
-            row.retail,
-            rates
-          )
-          row.currency = targetCurrency
-          return row
-        }),
-        next: data.next,
-        prev: data.prev,
-        count: data.count
+      // console.log(`rates`,rates)
```
Comment (blackmacaroon) [inline currency/index.js:83]: why do i need it or delete it

File: developer/developer-router.js [hunk 13]
```diff
@@ -3,45 +3,55 @@ const validate = require('../middleware/validate.js')
 const Developer = require('./developer-model.js')
 const router = express.Router()
 const convertCurrencies = require('../currency')
+const allowedPeriodFilter = require('../time-filter')
+
+// { apiCount: parseInt(req.count) }
```
Comment (blackmacaroon) [inline developer/developer-router.js]: souvenir?

File: developer/developer-router.js [hunk 14]
```diff
@@ -51,28 +61,37 @@ router.get(
...
             .then(converted => {
-              res.status(200).json({
-                warning: converted.warning,
-                message: req.message,
-                records: converted.data,
-                ratesUpdated: converted.ratesUpdated
+              // console.log(`converted: `,converted)
+              allowedPeriodFilter(converted,req.allowableTimePeriod)
+              .then(filtered => {
+                // console.log(filtered)
```
Comment (blackmacaroon) [inline developer/developer-router.js]: yeet

File: developer/developer-router.js [hunk 15]
```diff
@@ -51,28 +61,37 @@ router.get(
...
         } else {
-          convertCurrencies(records[0], req.currency) // Sauti wishes for all currency values to pass through conversion. See further notes in /currency
+          convertCurrencies(result, req.currency) // Sauti wishes for all currency values to pass through conversion. See further notes in /currency
             .then(converted => {
-              res.status(200).json({
-                warning: converted.warning,
-                message: req.message,
-                records: converted.data,
-                ratesUpdated: converted.ratesUpdated
+              // console.log(`converted: `,converted)
```
Comment (blackmacaroon) [inline developer/developer-router.js]: yeet

File: developer/developer-router.js [hunk 16]
```diff
@@ -126,14 +147,19 @@ router.get(
...
       console.log(error)
       res.status(500).send(error.message)
     })
 })
 
+
+//! requires filtering
 //Req.query needs product,startDate,endDate and returns a range of records
 //startDate is older than endDate
 //requires further validation possibly with moment.js to validate the date values//stretch goal for later
```
Comment (blackmacaroon) [inline developer/developer-router.js:156]: lets note this in the Notion Doc - Possible Future Stuff and Things

File: middleware/time-period-middleware.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+module.exports = async (req,res,next) => {
+
+    const { role } = req
+
+    // console.log(`time period middleware`,req)
```
Comment (blackmacaroon) [inline middleware/time-period-middleware.js:5]: yeet

File: developer/developer-router.js [hunk 18]
```diff
@@ -145,27 +171,33 @@ router.get(
   (req, res) => {
     Developer.getProductPriceRange(req.query)
       .then(records => {
+        console.log(`records product range `,records)
```
Comment (Sauti-Africa) [inline developer/developer-router.js]: is this necessary or for testing?

**Review by Sauti-Africa — CHANGES_REQUESTED** [review]
A couple ToDo notes, lets transfer these to trello/notion (with file name and line #) and clean up and testing console.logs before we merge to master.
Are we using the middleware/bottle.js or is that a remnant from Igor? Lets delete whole dern file.

**Pull request merged**